shared/lib_battery_lifetime_nmc.cpp

```diff
@@ -0,0 +1,216 @@
+    state->nmc_state->dq_relative_li_old = dq_new;
+    state->nmc_state->q_relative_li = (1.07 - (dq_new)) * 100;
+    return state->nmc_state->q_relative_li;
+}
+
+double lifetime_nmc_t::runQneg(double T_battery, double SOC) {
+
+    int dn_cycles = state->n_cycles - state->nmc_state->n_cycles_prev_day;
+
+    double c2 = state->nmc_state->c2_dt;
+    state->nmc_state->c2_dt = 0;
+
+    double dq_new;
+    if (state->nmc_state->dq_relative_neg_old == 0)
+        dq_new = 1 - sqrt(1 - 2 * (c2 / c0_ref) * dn_cycles);
+    else
+        dq_new = 1 - sqrt(1 - 2 * (c2 / c0_ref) * dn_cycles) + state->nmc_state->dq_relative_neg_old;
+
+    state->nmc_state->dq_relative_neg_old = dq_new;
+
+    state->nmc_state->q_relative_neg = (1 - (dq_new)) * 100;
+
+    //return state->nmc_state->q_relative_neg;
+    return 100;
+}
+
+void lifetime_nmc_t::runLifetimeModels(size_t lifetimeIndex, bool charge_changed, double prev_DOD, double DOD,
+                                       double T_battery) {
+    double q_last = state->q_relative;
+
+    // update day age of battery
+    size_t day_age_of_battery_old = (size_t)(state->day_age_of_battery);
+    state->day_age_of_battery += params->dt_hr / (double)util::hours_per_day;
+    auto ts_per_day = (size_t)(util::hours_per_day / params->dt_hr);
+
+    // convert battery temperature to Kelvin
+    T_battery += 273;
+    if (charge_changed)
+        cycle_model->rainflow(prev_DOD);
+
+    // update DOD_max if DOD > old DOD_max
+    if (DOD > state->nmc_state->DOD_max)
+        state->nmc_state->DOD_max = DOD;
+
+    //compute open circuit and negative electrode voltage as function of SOC
+    double SOC = 0.01 * (100 - DOD);
+    double DOD_max = state->nmc_state->DOD_max * 0.01;
+    double U_neg = calculate_Uneg(SOC);
+    double V_oc = calculate_Voc(SOC);
+
+    // compute lifetime degradation coefficients for current time step,
+    //multiply by timestep in days and populate corresponding vectors
+    double dt_day = (1. / 24) * params->dt_hr;
+    double b1_dt_el = b1_ref * exp(-(Ea_b_1 / Rug) * (1. / T_battery - 1. / T_ref))
+        * exp((alpha_a_b1 * F / Rug) * (U_neg / T_battery - U_ref / T_ref))
+        * exp(gamma * pow(DOD_max, beta_b1)) * dt_day;
+    double b2_dt_el = b2_ref * exp(-(Ea_b_2 / Rug) * (1. / T_battery - 1. / T_ref)) * dt_day;
+    double b3_dt_el = b3_ref * exp(-(Ea_b_3 / Rug) * (1. / T_battery - 1. / T_ref))
+        * exp((alpha_a_b3 * F / Rug) * (V_oc / T_battery - V_ref / T_ref))
+        * (1 + theta * DOD_max);
+    state->nmc_state->b1_dt += b1_dt_el;
+    state->nmc_state->b2_dt += b2_dt_el;
+    state->nmc_state->b3_dt += b3_dt_el;
+
+    //computations for q_neg
+    double c2_dt_el = c2_ref * exp(-(Ea_c_2 / Rug) * (1. / T_battery - 1. / T_ref))
+        * pow(0.01 * DOD, beta_c2);
+    state->nmc_state->c2_dt += c2_dt_el;
+
+    //Run capacity degradation model after every 24 hours
+    if (lifetimeIndex % ts_per_day == 23) {
```

Comment:
Has this been tested subhourly? My hunch is this could run 4 times per day with 15 minute code.

Reply:
Yes, good catch. I've fixed this to be `lifetimeIndex % ts_per_day == ts_per_day - 1`. Adding in subhourly tests